File: FocalStat.R
#!/usr/bin/env Rscript
library(exifr)
library(ggplot2)

# Function: Calculate adjusted focal length
process_images <- function(directory) {
  # Get all jpg files
  files <- list.files(directory, pattern = "\\.jpg$", recursive = TRUE, full.names = TRUE, ignore.case = TRUE)
  
  if (length(files) == 0) {
    message("No valid JPG files found")
    return(NULL)
  }
  
  # Extract EXIF data
  exif_data <- read_exif(files)
  
  # Filter relevant data
  if (!("FocalLength" %in% colnames(exif_data)) || 
      !("ImageWidth" %in% colnames(exif_data)) || 
      !("ImageHeight" %in% colnames(exif_data))) {
    message("Incomplete metadata; unable to extract focal length or dimensions")
    return(NULL)
  }
  
  # Calculate total pixels and adjusted focal length
  exif_data <- exif_data[!is.na(exif_data$FocalLength) & 
                           !is.na(exif_data$ImageWidth) & 
                           !is.na(exif_data$ImageHeight), ]
  exif_data$TotalPixels <- exif_data$ImageWidth * exif_data$ImageHeight
  exif_data$AdjustedFocalLength <- as.integer( exif_data$FocalLength/(exif_data$TotalPixels / 45441024) )
  
  return(exif_data$AdjustedFocalLength)
}

# Function: Plot histogram
plot_histogram <- function(data, output_file = "histogram.png") {
  if (is.null(data) || length(data) == 0) {
    message("No data available for plotting")
    return()
  }
  
  # Plot histogram
  p <- ggplot(data = data.frame(FocalLength = data), aes(x = FocalLength)) +
    geom_histogram(bins = 20, fill = "blue", color = "black", alpha = 0.7) +
    labs(title = "Adjusted Focal Length Distribution", x = "Adjusted Focal Length", y = "Frequency") +
    theme_minimal() +
    scale_x_continuous(breaks = c(10, 16, 24, 35, 50, 85, 105, 120, 200)) # Custom x-axis ticks
  
  # Save histogram with specified dimensions (width = 7, height = 5)
  ggsave(output_file, plot = p, width = 7, height = 5, units = "in")
  message("Histogram saved as ", output_file)
}

# Main program
args <- commandArgs(trailingOnly = TRUE)
if (length(args) == 0) {
  stop("Please provide a directory path as an argument. Example: Rscript process_images.R /path/to/directory")
}

directory <- args[1]
if (!dir.exists(directory)) {
  stop("Invalid directory path: ", directory)
}

data <- process_images(directory)

if (!is.null(data)) {
  plot_histogram(data)
} else {
  message("Processing complete, but no sufficient data for generating a histogram")
}

File: FocalStat.py
#!/usr/bin/env python3
import os
import exifread
from PIL import Image
import matplotlib.pyplot as plt
import sys
import numpy as np

Image.MAX_IMAGE_PIXELS = None  
def get_focal_length_and_pixels(image_path):
    try:
        with open(image_path, 'rb') as f:
            tags = exifread.process_file(f)
            
        focal_length = None
        if 'EXIF FocalLength' in tags:
            focal_length = tags['EXIF FocalLength']
            
        img = Image.open(image_path)
        width, height = img.size
        total_pixels = width * height
        
        if focal_length:
            focal_length = float(focal_length.values[0])  # EXIF返回的是一个比例
            coefficient = total_pixels / 45441024
            result = int( focal_length/coefficient)
            return result
        else:
            return None
    except Exception as e:
        print(f"can not process {image_path}: {e}")
        return None
    
def traverse_directory(directory):
    focal_lengths = []
    
    for root, dirs, files in os.walk(directory):
        print(f"processing directory: {root}")
        print(f"file: {files}")
        
        for file in files:
            # 检查是否是 .jpg 或 .jpeg 文件
            if file.lower().endswith(('.jpg', '.jpeg')):
                image_path = os.path.join(root, file)
                print(f"processing: {image_path}")
                result = get_focal_length_and_pixels(image_path)
                if result is not None:
                    focal_lengths.append(result)
                    
    return focal_lengths

import numpy as np
import matplotlib.pyplot as plt

def plot_histogram(focal_lengths):
    min_val = min(focal_lengths)
    max_val = max(focal_lengths)
    

    plt.hist(focal_lengths, bins=30, edgecolor='black') 
    plt.title('Histogram')
    plt.xlabel('Adjusted Focal Length')
    plt.ylabel('Frequency')
    
    custom_ticks = [10, 16, 24, 35, 50, 85, 105, 120, 200]
    plt.xticks(custom_ticks)  # 设置刻度为自定义的值
    plt.show()
    
    
def main():
    if len(sys.argv) != 2:
        print("Provide a valid directory")
        sys.exit(1)
        
    directory = sys.argv[1]
    focal_lengths = traverse_directory(directory)
    
    if focal_lengths:
        plot_histogram(focal_lengths)
    else:
        print("No .jpg found")
        
if __name__ == '__main__':
    main()
    